src/app/loadout-drawer/loadout-drawer-reducer.ts

```diff
@@ -1,363 +1,302 @@
+              ...newLoadout.parameters,
+              mods: newMods,
+            };
+            return { ...state, loadout: newLoadout };
+          }
         }
+        return state;
       }
-      return state;
     }
-  }
-
-  /**
-   * Produce a new loadout that adds a new item to the given loadout.
-   */
-  function addItem(
-    loadout: Readonly<Loadout>,
-    item: DimItem,
-    items: ResolvedLoadoutItem[],
-    equip?: boolean,
-    socketOverrides?: SocketOverrides
-  ): Loadout {
-    const loadoutItem: LoadoutItem = {
-      id: item.id,
-      hash: item.hash,
-      amount: 1,
-      equip: false,
-    };
-
-    // TODO: maybe we should just switch back to storing loadout items in memory by bucket
-
-    // Other items of the same type (as DimItem)
-    const typeInventory = items.filter((li) => li.item.bucket.hash === item.bucket.hash);
-    const dupe = loadout.items.find((i) => i.hash === item.hash && i.id === item.id);
-    const maxSlots = item.bucket.capacity;
-
-    return produce(loadout, (draftLoadout) => {
-      const findItem = ({ loadoutItem }: ResolvedLoadoutItem) =>
-        draftLoadout.items.find((i) => i.id === loadoutItem.id && i.hash === loadoutItem.hash)!;
-
-      if (!dupe) {
-        if (typeInventory.length < maxSlots) {
-          loadoutItem.equip =
-            equip !== undefined ? equip : item.equipment && typeInventory.length === 0;
-          if (loadoutItem.equip) {
-            for (const otherItem of typeInventory) {
-              findItem(otherItem).equip = false;
-            }
-          }
+  };
+}
 
-          // Only allow one subclass to be present per class (to allow for making a loadout that specifies a subclass for each class)
-          if (item.bucket.hash === BucketHashes.Subclass) {
-            const conflictingItem = items.find(
-              (li) =>
-                li.item.bucket.hash === item.bucket.hash && li.item.classType === item.classType
-            );
-            if (conflictingItem) {
-              draftLoadout.items = draftLoadout.items.filter(
-                (i) => i.id !== conflictingItem.item.id
-              );
-            }
-            loadoutItem.equip = true;
-          }
+/**
+ * Produce a new loadout that adds a new item to the given loadout.
+ */
+function addItem(
+  defs: D2ManifestDefinitions | D1ManifestDefinitions,
+  loadout: Readonly<Loadout>,
+  item: DimItem,
+  equip?: boolean,
+  socketOverrides?: SocketOverrides
+): Loadout {
+  const loadoutItem: LoadoutItem = {
+    id: item.id,
+    hash: item.hash,
+    amount: 1,
+    equip: false,
+  };
+  if (socketOverrides) {
+    loadoutItem.socketOverrides = socketOverrides;
+  }
 
-          if (socketOverrides) {
-            loadoutItem.socketOverrides = socketOverrides;
-          }
+  // TODO: We really want to be operating against the resolved items, right? Should we re-resolve them here, or what?
+  //       If we don't, we may not properly detect a dupe?
```

Comment:
I guess the two main options are:

1. Keep the resolved loadout around and check that for dupes (this can deal with arbitrarily powerful resolution logic).
2. Only implement a check that two `LoadoutItem`s resolve to the same item (which restricts what loadout item resolution can do to pretty much only doing any magic with `notransfer` buckets).

Reply:
I like that second idea!